hedera-node/src/main/java/com/hedera/services/contracts/execution/traceability/HederaTracer.java

```diff
@@ -0,0 +1,178 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.services.contracts.execution.traceability;
+
+import static com.hedera.services.contracts.execution.traceability.ContractActionType.CALL;
+import static com.hedera.services.contracts.execution.traceability.ContractActionType.CREATE;
+import static com.hedera.services.contracts.operation.HederaExceptionalHaltReason.INVALID_SOLIDITY_ADDRESS;
+
+import com.hedera.services.state.submerkle.EntityId;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.hyperledger.besu.evm.Code;
+import org.hyperledger.besu.evm.frame.ExceptionalHaltReason;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.frame.MessageFrame.State;
+
+@Singleton
+public class HederaTracer implements HederaOperationTracer {
+
+    private final List<SolidityAction> allActions;
+    private final Deque<SolidityAction> currentActionsStack;
+    private boolean areActionSidecarsEnabled;
+
+    @Inject
+    public HederaTracer() {
+        this.currentActionsStack = new ArrayDeque<>();
+        this.allActions = new ArrayList<>();
+    }
+
+    @Override
+    public void reset(final boolean areActionSidecarsEnabled) {
+        this.currentActionsStack.clear();
+        this.allActions.clear();
+        this.areActionSidecarsEnabled = areActionSidecarsEnabled;
+    }
+
+    @Override
+    public void traceExecution(MessageFrame frame, ExecuteOperation executeOperation) {
+        if (areActionSidecarsEnabled && currentActionsStack.isEmpty()) {
+            trackActionFor(frame, 0);
+        }
+
+        executeOperation.execute();
+
+        if (areActionSidecarsEnabled) {
+            final var frameState = frame.getState();
+            if (frameState != State.CODE_EXECUTING) {
+                if (frameState == State.CODE_SUSPENDED) {
+                    final var nextFrame = frame.getMessageFrameStack().peek();
+                    trackActionFor(nextFrame, nextFrame.getMessageFrameStack().size() - 1);
+                } else {
+                    finalizeActionFor(currentActionsStack.pop(), frame, frameState);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void tracePrecompileResult(final MessageFrame frame, final ContractActionType type) {
+        if (areActionSidecarsEnabled) {
+            final var lastAction = currentActionsStack.pop();
+            // specialize the call type - precompile or system call
+            lastAction.setCallType(type);
+            // we have to null out recipient account and set recipient contract
```

Comment:
These 2 along with most of the comments (e.g. every comment that begins with the word 'set') in this file are completely omittable. If the info given by the comment can be inferred from the line of code the comment becomes redundant.

Reply:
Agreed, some of the comments just mirrored the code --- removed now.